File: go/src/RumbleRacing/core/asset/txf/ZTHE.go
package txf

import (
	"encoding/binary"
)

type ZTHE struct {
	TextureCount uint32
	Textures     []ZTHETexture

	RawData []byte
}

type ZTHETexture struct {
	Images             []ZTHETextureMetaHeader
	TexelStorageFormat uint8
	ImageCount         uint8 // Each texture could have up to 4 mip-mapped sub-textures
	BlockWidthPixels   uint16
	CLUTHeaderIndex    uint8
	TextureId          uint16 // Potential texture ID that 3d model looks up?

	RawData []byte
}

type ZTHETextureMetaHeader struct {
	TXDAAddressOffset   uint32
	Unk                 uint16
	BlockHeightPixels   uint16
	SelfPlusMemAllocRes uint16
	RamDestWidth        uint16
}

func parseZTHE(buf []byte) (*ZTHE, error) {

	// size := binary.LittleEndian.Uint32(buf[4:8])
	// fmt.Println(size)
	raw := buf

	texCount := binary.LittleEndian.Uint32(buf[8 : 8+4])
	buf = buf[8+4:]

	var textures []ZTHETexture

	for i := 0; i+0x48 <= len(buf); i += 0x48 {
		data := buf[i : i+0x48]
		rawTexture := data
		// fmt.Println(hex.Dump(data[0:8]))

		var metaHeaders []ZTHETextureMetaHeader

		imageCount := data[0x31]

		for j := 0; j < int(imageCount); j++ {
			offset := j * 0xc
			hData := data[offset : offset+0xc]
			// fmt.Println("meta:", hex.Dump(hData))

			metaHeaders = append(metaHeaders, ZTHETextureMetaHeader{
				TXDAAddressOffset:   binary.LittleEndian.Uint32(hData[0:4]),
				Unk:                 binary.LittleEndian.Uint16(hData[0x4 : 0x4+2]),
				BlockHeightPixels:   binary.LittleEndian.Uint16(hData[0x6 : 0x6+2]),
				SelfPlusMemAllocRes: binary.LittleEndian.Uint16(hData[0x8 : 0x8+2]),
				RamDestWidth:        binary.LittleEndian.Uint16(hData[0xa : 0xa+2]),
			})
		}

		textures = append(textures, ZTHETexture{
			TexelStorageFormat: data[0x30],
			ImageCount:         imageCount,
			BlockWidthPixels:   binary.LittleEndian.Uint16(data[0x3e:(0x3e + 2)]),
			Images:             metaHeaders,
			TextureId:          binary.LittleEndian.Uint16(data[0x34:(0x34 + 2)]),
			CLUTHeaderIndex:    data[0x44],
			RawData:            rawTexture,
		})
	}

	if int(texCount) != len(textures) {
		panic("TexCount != length of textures!")
	}

	return &ZTHE{
		TextureCount: texCount,
		Textures:     textures,
		RawData:      raw,
	}, nil
}

File: src/RumbleRacing/asset/txf/ZTHE.ts
import ArrayBufferSlice from "../../../ArrayBufferSlice";
import { GSPixelStorageFormat } from "../../../Common/PS2/GS";

export interface ZTHETextureMetaHeader {
  txdaAddressOffset: number;
  blockHeightPixels: number;
  selfPlusMemAllocRes: number;
  ramDestWidth: number;
}

export interface ZTHETexture {
  images: ZTHETextureMetaHeader[];
  texelStorageFormat: GSPixelStorageFormat;
  imageCount: number;
  blockWidthPixels: number;
  clutHeaderIndex: number;
  textureId: number;
  rawData: ArrayBufferSlice;
}

export interface ZTHE {
  textureCount: number;
  textures: ZTHETexture[];
  rawData: ArrayBufferSlice;
}

export function parseZTHE(buf: ArrayBufferSlice): ZTHE {
  const texCount = buf.createDataView().getUint32(8, true);
  const body = buf.subarray(12);

  const textures: ZTHETexture[] = [];

  for (let i = 0; i + 0x48 <= body.byteLength; i += 0x48) {
    const data = body.subarray(i, 0x48);
    const dataView = data.createDataView();
    const imageCount = dataView.getUint8(0x31);

    const metaHeaders: ZTHETextureMetaHeader[] = [];
    for (let j = 0; j < imageCount; j++) {
      const offset = j * 0xc;
      metaHeaders.push({
        txdaAddressOffset: dataView.getUint32(offset, true),
        blockHeightPixels: dataView.getUint16(offset + 0x6, true),
        selfPlusMemAllocRes: dataView.getUint16(offset + 0x8, true),
        ramDestWidth: dataView.getUint16(offset + 0xa, true),
      });
    }

    textures.push({
      texelStorageFormat: dataView.getUint8(0x30),
      imageCount,
      blockWidthPixels: dataView.getUint16(0x3e, true),
      images: metaHeaders,
      textureId: dataView.getUint16(0x34, true),
      clutHeaderIndex: dataView.getUint8(0x44),
      rawData: data,
    });
  }

  if (texCount !== textures.length) {
    throw new Error("TexCount != length of textures!");
  }

  return { textureCount: texCount, textures, rawData: buf };
}
